### ISSUE TITLE
Very low accuracy in the mnist_cnn when running on a GPU using tensorflow  backend

### ISSUE THREAD

**Issue by vivounicorn** [fchollet/keras#3508]
hi everyone:
   I tried all the examples of MNIST and when i use cpu all of them are work well (tensorflow and theano all tried), but **when i use gpu (tensorflow  backend) mnist_cnn.py is not working, it has very low accuracy**, i guess there's some wrong with cnn on gpu.
([issue 511](https://github.com/fchollet/keras/issues/511) can not solve the problem.)

### 1、.keras/keras.json
{
    "image_dim_ordering": "tf",
    "epsilon": 1e-05,
    "floatx": "float32",
    "backend": "tensorflow"
}

### 2、source code is changed like this：
line 35：X_train = X_train.reshape(X_train.shape[0], img_rows, img_cols, 1)
line 36：X_test = X_test.reshape(X_test.shape[0], img_rows, img_cols, 1)
line 51：model.add(Convolution2D(nb_filters, nb_conv, nb_conv,
line 52：                          border_mode='valid',
line 53：                          input_shape=(img_rows, img_cols, 1)))

### 1、on gpu
(tensorflow):~/dllearning/mnist$ python mnist_cnn.py 
I tensorflow/stream_executor/dso_loader.cc:108] successfully opened CUDA library libcublas.so locally
I tensorflow/stream_executor/dso_loader.cc:108] successfully opened CUDA library libcudnn.so locally
I tensorflow/stream_executor/dso_loader.cc:108] successfully opened CUDA library libcufft.so locally
I tensorflow/stream_executor/dso_loader.cc:108] successfully opened CUDA library libcuda.so locally
I tensorflow/stream_executor/dso_loader.cc:108] successfully opened CUDA library libcurand.so locally
Using TensorFlow backend.
X_train shape: (60000, 28, 28, 1)
60000 train samples
10000 test samples
I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:924] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
I tensorflow/core/common_runtime/gpu/gpu_init.cc:102] Found device 0 with properties: 
name: GeForce GTX 1070
major: 6 minor: 1 memoryClockRate (GHz) 1.7845
pciBusID 0000:01:00.0
Total memory: 7.92GiB
Free memory: 1.93GiB
I tensorflow/core/common_runtime/gpu/gpu_init.cc:126] DMA: 0 
I tensorflow/core/common_runtime/gpu/gpu_init.cc:136] 0:   Y 
I tensorflow/core/common_runtime/gpu/gpu_device.cc:806] Creating TensorFlow device (/gpu:0) -> (device: 0, name: GeForce GTX 1070, pci bus id: 0000:01:00.0)
Train on 60000 samples, validate on 10000 samples
Epoch 1/12
60000/60000 [==============================] - 12s - loss: 2.3065 - acc: 0.1103 - val_loss: 2.3011 - val_acc: 0.1137
Epoch 2/12
60000/60000 [==============================] - 6s - loss: 2.3022 - acc: 0.1122 - val_loss: 2.3011 - val_acc: 0.1135
Epoch 3/12
60000/60000 [==============================] - 6s - loss: 2.3014 - acc: 0.1123 - val_loss: 2.3011 - val_acc: 0.1135
Epoch 4/12
60000/60000 [==============================] - 6s - loss: 2.3025 - acc: 0.1121 - val_loss: 2.3010 - val_acc: 0.1135
Epoch 5/12
60000/60000 [==============================] - 5s - loss: 2.3013 - acc: 0.1124 - val_loss: 2.3010 - val_acc: 0.1135
Epoch 6/12
60000/60000 [==============================] - 5s - loss: 2.3013 - acc: 0.1124 - val_loss: 2.3010 - val_acc: 0.1135
Epoch 7/12
60000/60000 [==============================] - 6s - loss: 2.3013 - acc: 0.1124 - val_loss: 2.3021 - val_acc: 0.1135
Epoch 8/12
60000/60000 [==============================] - 6s - loss: 2.3013 - acc: 0.1124 - val_loss: 2.3017 - val_acc: 0.1135
Epoch 9/12
60000/60000 [==============================] - 6s - loss: 2.3013 - acc: 0.1123 - val_loss: 2.3010 - val_acc: 0.1135
Epoch 10/12
60000/60000 [==============================] - 5s - loss: 2.3012 - acc: 0.1123 - val_loss: 2.3011 - val_acc: 0.1135
Epoch 11/12
60000/60000 [==============================] - 6s - loss: 2.3013 - acc: 0.1124 - val_loss: 2.3011 - val_acc: 0.1135
Epoch 12/12
60000/60000 [==============================] - 6s - loss: 2.3013 - acc: 0.1124 - val_loss: 2.3011 - val_acc: 0.1135
Test score: 2.30111371231
Test accuracy: 0.1135

### 2、on cpu
Train on 60000 samples, validate on 10000 samples
Epoch 1/12
60000/60000 [==============================] - 71s - loss: 0.3994 - acc: 0.8764 - val_loss: 0.1060 - val_acc: 0.9674
Epoch 2/12
60000/60000 [==============================] - 72s - loss: 0.1495 - acc: 0.9557 - val_loss: 0.0677 - val_acc: 0.9785
Epoch 3/12
60000/60000 [==============================] - 72s - loss: 0.1119 - acc: 0.9675 - val_loss: 0.0537 - val_acc: 0.9823
Epoch 4/12
60000/60000 [==============================] - 100s - loss: 0.0914 - acc: 0.9724 - val_loss: 0.0474 - val_acc: 0.9839
Epoch 5/12
60000/60000 [==============================] - 114s - loss: 0.0805 - acc: 0.9758 - val_loss: 0.0393 - val_acc: 0.9874
Epoch 6/12
60000/60000 [==============================] - 117s - loss: 0.0721 - acc: 0.9783 - val_loss: 0.0381 - val_acc: 0.9873
Epoch 7/12
60000/60000 [==============================] - 118s - loss: 0.0664 - acc: 0.9806 - val_loss: 0.0376 - val_acc: 0.9874
Epoch 8/12
60000/60000 [==============================] - 121s - loss: 0.0621 - acc: 0.9815 - val_loss: 0.0338 - val_acc: 0.9883
Epoch 9/12
60000/60000 [==============================] - 121s - loss: 0.0565 - acc: 0.9834 - val_loss: 0.0322 - val_acc: 0.9887
Epoch 10/12
60000/60000 [==============================] - 124s - loss: 0.0542 - acc: 0.9840 - val_loss: 0.0304 - val_acc: 0.9906
Epoch 11/12
60000/60000 [==============================] - 121s - loss: 0.0491 - acc: 0.9850 - val_loss: 0.0314 - val_acc: 0.9896
Epoch 12/12
60000/60000 [==============================] - 116s - loss: 0.0485 - acc: 0.9857 - val_loss: 0.0294 - val_acc: 0.9898
Test score: 0.0294442383148
Test accuracy: 0.9898

**Comment by vivounicorn** [comment]
May be it's my tensorflow‘s problem because the following code can not work well too：
``` python
import tensorflow as tf
from tensorflow.examples.tutorials.mnist import input_data

def weight_varible(shape):
    initial = tf.truncated_normal(shape, stddev=0.1)
    return tf.Variable(initial)

def bias_variable(shape):
    initial = tf.constant(0.1, shape=shape)
    return tf.Variable(initial)

def conv2d(x, W):
    return tf.nn.conv2d(x, W, strides=[1, 1, 1, 1], padding='SAME')

def max_pool_2x2(x):
    return tf.nn.max_pool(x, ksize=[1, 2, 2, 1], strides=[1, 2, 2, 1], padding='SAME')


mnist = input_data.read_data_sets("MNIST_data/", one_hot=True)
print("Download Done!")

sess = tf.InteractiveSession()

# paras
W_conv1 = weight_varible([5, 5, 1, 32])
b_conv1 = bias_variable([32])

# conv layer-1
x = tf.placeholder(tf.float32, [None, 784])
x_image = tf.reshape(x, [-1, 28, 28, 1])

h_conv1 = tf.nn.relu(conv2d(x_image, W_conv1) + b_conv1)
h_pool1 = max_pool_2x2(h_conv1)

# conv layer-2
W_conv2 = weight_varible([5, 5, 32, 64])
b_conv2 = bias_variable([64])

h_conv2 = tf.nn.relu(conv2d(h_pool1, W_conv2) + b_conv2)
h_pool2 = max_pool_2x2(h_conv2)

# full connection
W_fc1 = weight_varible([7 * 7 * 64, 1024])
b_fc1 = bias_variable([1024])

h_pool2_flat = tf.reshape(h_pool2, [-1, 7 * 7 * 64])
h_fc1 = tf.nn.relu(tf.matmul(h_pool2_flat, W_fc1) + b_fc1)

# dropout
keep_prob = tf.placeholder(tf.float32)
h_fc1_drop = tf.nn.dropout(h_fc1, keep_prob)

# output layer: softmax
W_fc2 = weight_varible([1024, 10])
b_fc2 = bias_variable([10])

y_conv = tf.nn.softmax(tf.matmul(h_fc1_drop, W_fc2) + b_fc2)
y_ = tf.placeholder(tf.float32, [None, 10])

# model training
cross_entropy = -tf.reduce_sum(y_ * tf.log(y_conv))
train_step = tf.train.AdamOptimizer(1e-4).minimize(cross_entropy)

correct_prediction = tf.equal(tf.arg_max(y_conv, 1), tf.arg_max(y_, 1))
accuracy = tf.reduce_mean(tf.cast(correct_prediction, tf.float32))

sess.run(tf.initialize_all_variables())

for i in range(20000):
    batch = mnist.train.next_batch(50)

    if i % 100 == 0:
        train_accuacy = accuracy.eval(feed_dict={x: batch[0], y_: batch[1], keep_prob: 1.0})
        print("step %d, training accuracy %g"%(i, train_accuacy))
    train_step.run(feed_dict = {x: batch[0], y_: batch[1], keep_prob: 0.5})

# accuacy on test
print("test accuracy %g"%(accuracy.eval(feed_dict={x: mnist.test.images, y_: mnist.test.labels, keep_prob: 1.0})))
```

**Comment by vivounicorn** [comment]
@DominicBreuker tks

**Comment by vivounicorn** [comment]
[https://github.com/tensorflow/tensorflow/issues/3891](url)

**Comment by jinchenglee** [comment]
I have similar problem with tensorflow with gtx1070, ubuntu 16.04. I'm trying this: http://abhay.harpale.net/blog/machine-learning/deep-learning/getting-tensorflow-to-work-with-gpu-nvidia-gtx-1080-on-ubuntu-16-04-lts/

**Comment by vivounicorn** [comment]
@jinchenglee 3ks
